### ISSUE TITLE
[Feature] Tuning / per-GPU config for DSv4 top-k v2 cluster launch plan (kClusterFloor / kNumPersistentClusters / kCandidates)

### ISSUE THREAD

**Issue by dahaobenhao** [sgl-project/sglang#37150]
## Summary

The DSv4 top-k v2 cooperative-cluster kernel (`python/sglang/kernels/jit/csrc/deepseek_v4/topk_v2.cuh`) selects its launch plan from **hardcoded, B200/sm100-derived `constexpr` values**. There is no benchmark harness and no per-GPU tuned-config / autotune path for these parameters, even though the code itself flags the plan as tunable. This is the one piece of the DSA decode path that has no tuning story, in contrast to fused-MoE and DeepGEMM which both do.

## The hardcoded launch plan (main)

In `topk_v2.cuh`:

```cpp
constexpr uint32_t kClusterFloor = 65536;                 // L58
constexpr uint32_t kClusterMaxBatch = 512;                // L59
constexpr uint32_t kNumPersistentClusters = 15 * kOccupancy;  // L60  (15 == B200 one-wave cluster count)
...
// L379: "cap = pool size only fits the shortest (~98K, one-wave) bucket. (Plan is tunable.)"
constexpr Pair kCandidates[] = { {65536,30},{98304,48},{131072,60},{196608,80},{262144,112},{393216,128} };  // L384
...
constexpr uint32_t kClusterFloorSmall = 32768;            // L545
constexpr uint32_t kSmallBatchLowFloor = 15;              // L546
```

`kNumPersistentClusters = 15 * kOccupancy` is literally the B200 single-wave cluster count; the `kCandidates` `{threshold, max_batch_size}` plan table and the two cluster floors are B200-tuned magic numbers. The `(Plan is tunable.)` comment states the intent, but nothing exposes it — no env var, no config file, no autotune. On any non-B200 SM (different SM count / occupancy / L2 / cluster-launch behavior), these are heuristics, not tuned values.

#35095 (adaptive small-batch cluster split) improves the dispatch and explicitly notes the split caps are *"B200/sm100 autotune values scaled by SM count at runtime"* — i.e. still baked B200 constants scaled by a scalar, not values tuned for the target device. That PR makes the split adaptive; it does not give the plan a tuning mechanism.

## Why this matters (measured)

On B200 the plan is fine — decode top-k stays cheap. Measured decode share (v0.5.17, TP8, graph-off, compute-relative i.e. after removing the all-reduce spin floor):

| model | K | top-k @32K | top-k @1M | kernel |
|-------|---|-----------|-----------|--------|
| DeepSeek-V4-Flash | 512 | 1.4% | 4.1% | `topk_small_batch_kernel` |
| DeepSeek-V4-Pro | 1024 | 2.1% | 4.8% | `topk_small_batch_kernel` |

`topk_small_batch_kernel` (the cooperative `__cluster_dims__` kernel) is the only decode top-k path, and per-launch cost scales sub-linearly with context because the persistent-cluster pool is sized for the B200 wave. The concern is portability: the same constants on a device with a different SM/occupancy profile will either under-fill the pool (idle clusters) or oversubscribe it (extra waves), and there is currently no supported way to retune without editing and recompiling the `.cuh`.

## Asymmetry with existing tuning infra

- fused-MoE: `benchmark/kernels/fused_moe_triton/tuning_fused_moe_triton.py`, `tuning_client.py`, tuned config JSONs.
- DeepGEMM: `benchmark/kernels/deepseek/benchmark_deepgemm_fp8_gemm*.py`.
- DSA indexer scoring: `benchmark/kernels/deepseek/benchmark_cute_dsl_fp8_paged_mqa_logits.py`.
- **DSv4 top-k v2 cluster kernel: no benchmark, no tuning script, no per-device config.**

## Proposed feature

1. A `benchmark/kernels/deepseek/benchmark_topk_v2.py` that sweeps (batch, max_seq_len, K, page_size) and reports the cluster-vs-streaming crossover and pool occupancy — so the plan can be measured per device.
2. Make the launch plan device-tunable rather than a compile-time `constexpr`: either a small tuned-config lookup keyed by device name (mirroring the MoE config JSONs) selecting `{kClusterFloor, kClusterFloorSmall, kNumPersistentClusters, kCandidates}`, or a documented runtime override, with the current B200 values as the default entry.

Happy to contribute this; it slots on top of #35095 (whose split caps would become the first tuned entry) and is independent of the correctness issue #35257.

**Comment by dahaobenhao** [comment]
Correcting two numbers in the description above, and withdrawing one of my two proposals.
Everything else in the issue I have re-verified against `main` at `1fb85053e7` (2026-09-03)
and it stands.

**The DSv4-Pro percentages are wrong.** The table I quoted (2.1% at 32k, 4.8% at 1M) came
from a first-pass profile whose top-k bucket was selected by a `/topk/` regex. On the Pro
model that regex also swept in `_pack_topk_ids_triton_kernel` — 3521.8 us, n=1952=61x32,
which is MoE expert-id packing, not DSA top-k. Re-derived with an explicit DSA whitelist
(`topk_small_batch|topk_persistent|topk_plan`, launches 992 = 960 + 32 plan), same traces:

| model | K | 32k | 1M | as filed |
|---|---|---|---|---|
| DSv4-Flash | 512 | 1.405% | 4.102% | 1.4% / 4.1% — unaffected, no `_pack_topk_ids` on the fp8 MoE path |
| DSv4-Pro | 1024 | **1.315%** | **4.130%** | 2.1% / 4.8% — **retracted** |

Percent of compute (total minus comm), decode, v0.5.17, TP8, 8xB200, graph off. So the
headroom is smaller than I filed it, and the same on both models rather than larger on Pro.

**Withdrawing proposal (1) as written.** A benchmark for this kernel already exists:
`test/registered/kernels/benchmark/attention/bench_topk.py` imports `plan_topk_v2`,
`topk_transform_paged_v2` and `topk_transform_ragged_v2`, and is CI-registered
(`stage="base-b-kernel-benchmark"`). My "no benchmark or tuning script" sentence is simply
false and a new file under `benchmark/kernels/deepseek/` would duplicate it. What is
actually missing is narrower: that sweep is over `seq_len` 2^10..2^18 and `batch_size`
2^0..2^12 only — no `K` axis, no page-size axis, and it reports neither the
cluster-vs-streaming crossover nor persistent-pool occupancy. Extending it is the right
form of the ask, and it belongs to whoever adds provider lines there.

**What the issue still asks for, unchanged.** On `main` today the launch plan is
compile-time only — `kClusterFloor = 65536`, `kClusterMaxBatch = 512`,
`kNumPersistentClusters = 15 * kOccupancy`, the six-entry `kCandidates` threshold/batch
table, and the host-side `kClusterFloorSmall = 32768` / `kSmallBatchLowFloor = 15` pair
(introduced in #26788, whose comment states "B200 fits one wave of exactly 15 such
clusters"). `topk_v2.cuh` says "(Plan is tunable.)" and there is no `getenv`, no config
file, and no tuned-config lookup anywhere in `python/sglang/kernels/jit/csrc/deepseek_v4/`
that would let anyone tune it without editing the header and recompiling. #35095 (open)
does add per-device rescaling with `get_sm_count` / `get_l2_cache_size` for the split
factor it introduces, from sm_100 anchors — which is the shape I am asking for, applied to
the rest of the plan.

To be explicit about scope, since the corrected numbers weaken the motivation: at no more than 4.13%
of decode compute this is not a B200 performance complaint. It is portability — on a device
with a different SM count or occupancy profile the persistent pool is either partly idle or
oversubscribed, and today the only remedy is a header edit.
